### PR TITLE
GUIInfoManager: Move SetCurrentItem into a background job

### PR DESCRIPTION
**Pull request by popcornmix**
This often takes several hundred milliseconds on a Pi2 and can cause timeouts on renderer configure

### CODE REVIEW

**Comment by popcornmix** [comment]
To elaborate, this call does sql database accesses and is called from application thread so blocks renderer for a considerable time.

**Comment by FernetMenta** [comment]
+1

File: xbmc/GUIInfoManager.h [hunk 1]
```diff
@@ -152,6 +152,7 @@ class CGUIInfoManager : public IMsgTargetCallback, public Observable,
   std::string GetDuration(TIME_FORMAT format = TIME_FORMAT_GUESS) const;
 
   void SetCurrentItem(CFileItem &item);
+  void SetCurrentItemInternal(CFileItem &item);
```
Comment (Razzeee) [inline xbmc/GUIInfoManager.h]: Wouldn't it make sense to make this private if it is indeed internal?

Comment (popcornmix) [inline xbmc/GUIInfoManager.h]: It's called from JobManager so needs to be public.
Perhaps naming it SetCurrentItemBlocking or using the existing function with a blocking parameter would be better?

Comment (Razzeee) [inline xbmc/GUIInfoManager.h]: I would probably go for a parameter and some comments/doxygen comments.

Comment (Paxxi) [inline xbmc/GUIInfoManager.h]: Keeping it separate is the best way to go,  we have way too many methods that take Boolean parameters which requires docs and or looking at implementation to know what it does. Make it obvious when seeing the invocation that it runs off thread, like async, background, nonblocking, just pick something nice

**Comment by popcornmix** [comment]
Just updated with parameter before seeing Paxxi's comment.
Is general consensus to have a separate funtion?

**Comment by Razzeee** [comment]
I'm fine with whatever is understandable on first sight :)
So pick whatever you like or stay with this if it fits/your lazy ;)

**Comment by MrMC** [comment]
Does anything need a valid currentItem value before the job might complete ?

**Comment by popcornmix** [comment]
@MrMC : https://github.com/popcornmix/xbmc/commit/aa533ea2ff3e43a89bcb5896b91c5f28cebd0894#commitcomment-15263328

**Comment by MrMC** [comment]
haha, yea, that will surely fix it. I'd change the code to job out the database access and leave the runtime set of currentItem. The database is just status update on that item. currentItem seems important to have valid asap, but that's just me.

**Comment by FernetMenta** [comment]
@MrMC in a later stage I would agree but we are in early alpha and there is enough time for things to come to surface if something breaks. At this stage I would rather break it than hide issues.

**Comment by popcornmix** [comment]
jenkins build this please

**Comment by popcornmix** [comment]
Looks like this might be responsible for a crash: http://pastebin.com/xh5n1yPC

**Comment by koying** [comment]
An unfortunate side-effect of that one is that it breaks the assumption that the infobools are available at "OnPlay" annoucement time, which is a problem when you want to publish the current playing item.

I speak for myself, obviously, because I need it, but I'm pretty sure this breaks thumbnails on upnp as well (https://github.com/xbmc/xbmc/blob/69f4545bbfe147d169a35714a0cedc4ed0222f63/xbmc/network/upnp/UPnPRenderer.cpp#L412) or it works by chance. @Montellese ?

I could use the fileitem passed by "OnPlay", but that just means redoing what infomanager does, really, doubling the db accesses.

Any suggestions?
An additional "OnInfoAvailable" annoucement, maybe?

**Comment by FernetMenta** [comment]
All player events should be non blocking. Then players can fire the events at the right time. The event handler should take care that handling of one action has completed before next one is started.

**Comment by koying** [comment]
Not sure I understand you, but the issue is not blocking or not blocking, it is that that infobool updating is asynchronous vs. the firing of "OnPlay", so "OnPlay" receivers do not (or do not know) if they have the latest infobools.

I'd actually assume that it'd be an issue for addons like trakt also, that needs the latest info in the OnPlay event. 
Or is it working by chance, because the infobool updating always finish before the python events are fired?
Or am I missing something?

**Comment by FernetMenta** [comment]
Currently there are two issues:

1) OnPlay is fired too early at a time playback has not yet started. VP can't fire those events on its time critical path because those are blocking.

2) (Player) events are not handled by a single thread and as a result one event may overtake another one.

IPlayerCallback should not be implemented by CApplication that runs on main thread but some other instance dedicated to this purpose.

**Comment by koying** [comment]
I'm sure both your points are very valid, but probably not immediately related to the issue at hand ;)
The InfoManager updating is only partially related to the Player, really. It's the currently "playing" item *metadata* that is at hand, here.

Thus my idea to have an additional event fired when those metadatas are available, with probably a different category.

**Comment by FernetMenta** [comment]
Basically there are two options. Fix it properly by fixing the root cause or add another quick hack to the behavior that is already out of control because of too many design flaws.
Adding "OnInfoAvailable" belongs to the latter.

**Comment by FernetMenta** [comment]
@DaveTBlake this is the reason for the crashes becuase there are no locks. CApp calls ResetItem while the job sets the item.
instead of running SetItem on a job we need to offload the database operations from the main thread.

I'll revert this change here

**Comment by DaveTBlake** [comment]
Not sure reverting this is the best fix.
But I can comfirm that as I suspected the crash is a timing issue - CApp resets the item while the GUIInfoItem job is tryng to use it

**Comment by FernetMenta** [comment]
GUIInfoManager is not designed for multithreading. Hence the revert is best option. Introducing locks would be a hack and lead to original issue again becuase renderthread will be blocked during db operations.

**Comment by koying** [comment]
@mrmc suggestion seems to make sense: only offload the DB update while keeping the current item update on main thread, doesn't it?
Would solve both the lock and the crash

**Comment by FernetMenta** [comment]
right

**Comment by MrMC** [comment]
You are updating the database, not the item.  That's the lengthy part which stalls the main thread.

**Comment by DaveTBlake** [comment]
@MrMC I really only see SELECT statements, all gets no updates. Can tell me where SetCurrentItem is updating the db?

**Comment by popcornmix** [comment]
I think any database access is too slow for running in the main (also used by video renderer) thread.

**Comment by MrMC** [comment]
I might be wrong then. But the idea is to offload the db access if the current item does not need the info to startup playback. So far I don't think anyone has looked into exactly what it being updated and if playback startup requires such info. If not, move the update to a job.

**Comment by DaveTBlake** [comment]
That could well be true @popcornmix , I was just pointing out that unlike DB updates that could be fire-and-forget, this is reading and the item wants the results.

**Comment by MrMC** [comment]
The item might 'want' the results but does it 'need' the results to startup playback ?

**Comment by DaveTBlake** [comment]
This does not only happen at start of playback, but no there is no rush for GUI data AFAIK, the user can wait for it to show as long as rendering doesn't stop

**Pull request merged**